Automattic/woocommerce-payments
# Adds fixed and consistent naming pattern for Stripe Link payment tokens

Fixes #5022

#### Changes proposed in this Pull Request


# Resolving Stripe Link troubling token titles

It is possible to use Stripe Link as a saved payment method, however, the display of the token is traumatically twisted on the blocks checkout. Currently on both *My account > Payment methods* page and shortcode checkout, the token will be displayed with a label of the format of "Stripe Link email example@example.com", whereas on the blocks checkout the token will be labelled "Saved token for woocommerce_payments". This is particularly unhelpful, because it does not even indicate what type of payment method the token is for.

This is because the [default label for a saved payment method token in the blocks checkout](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/cart-checkout-shared/payment-methods/saved-payment-method-options.tsx#L67) is "Saved token for { payment gateway ID }" when the token is missing either a `brand` or `last4` property. This is definitely code that is more specific to card numbers, for which both of these properties is particularly meaningful, and seems incompatible and irrelevant to Stripe Link payment methods. However, I have found a compromise that I believe fits this format, looks clean, and most importantly somewhat makes sense. 

In this PR I update the label format to use "Stripe Link email" as the `brand` property and use a redacted representation of the last four characters of a customer's email handle for the `last4` property. This allows us to align Stripe Link with other payment tokens supported by WooCommerce and WooCommerce Blocks, while still remaining meaningful and presentable. Using the example above, a Link token using the customer email address of "example@example.com" will be transformed into the label "Stripe Link email ending in ***mple@example.com". This keeps this email property somewhat private while also preventing excessively long email handles from overruning in the checkout field, since this address handle will now contain a maximum of seven characters. 



## Before

![My accounts payment methods before](https://cdn-std.droplr.net/files/acc_1104206/fF5wsH)
_Payment methods on My accounts page display_

![Shortcode checkout payment token display](https://cdn-std.droplr.net/files/acc_1104206/tSsILU)
_Shortcode checkout display_

![Blocks checkout payment token display](https://cdn-std.droplr.net/files/acc_1104206/HhNB3M)
_Blocks checkout display_

## After

![My accounts payment methods before](https://cdn-std.droplr.net/files/acc_1104206/yBS9FP)
_Payment methods on My accounts page display_

![Shortcode checkout payment token display](https://cdn-std.droplr.net/files/acc_1104206/tiUCX0)
_Shortcode checkout display_

![Blocks checkout payment token display](https://cdn-std.droplr.net/files/acc_1104206/aLmwgm)
_Blocks checkout display_

#### Testing instructions


1. On `develop` complete a checkout using Link and save the payment method during checkout. You should find that the email address you used is present amongst the saved tokens on the shortcode checkout, but not the blocks checkout.
2. Change to this branch and notice that your saved Link token titles are transformed to the above described format.
3. Confirm that Link payment token titles are consistent between shortcode checkout, blocks checkout, and my account (*My account > Payment methods*) pages.
4. Save a new Link payment method and confirm that labels are still consistent and correct.


-------------------


**Post merge**
